### ISSUE TITLE
Additional requested embed players

### ISSUE THREAD

**Issue by pfrazee** [bluesky-social/social-app#2314]
- Vimeo
- Giphy
- Tenor
- Imgur
- TikTok
- Instagram (?)

We can anything that's secure and has moderation on their platform!

**Comment by seboslaw** [comment]
As a side-note. The app currently automatically opts-in users to this functionality. However, from a data privacy perspective this is kind of edgy since the app now per default (at least) leaks the user's IP address to 3rd parties like Google, Meta, etc. pp. 
So imho this feature should be opt-in and there should be a disclaimer informing the users what information is passed onto the 3rd parties when enabling the feature.

**Comment by pfrazee** [comment]
Another thing to add to the list, handle the youtube mobile links that prefix with `m.`

**Comment by pfrazee** [comment]
> As a side-note. The app currently automatically opts-in users to this functionality. However, from a data privacy perspective this is kind of edgy since the app now per default (at least) leaks the user's IP address to 3rd parties like Google, Meta, etc. pp.

It only contacts those services if the users tap into it and I think people generally understand that those services have to be contacted to display content from them

**Comment by seboslaw** [comment]
> > As a side-note. The app currently automatically opts-in users to this functionality. However, from a data privacy perspective this is kind of edgy since the app now per default (at least) leaks the user's IP address to 3rd parties like Google, Meta, etc. pp.
> 
> It only contacts those services if the users tap into it and I think people generally understand that those services have to be contacted to display content from them

It really depends on the market you're in. I'm coming from a European / GDPR centric POV where you'd have to have the user  opt-in to this feature, even if you're only running it when the user clicks on it. But that's only for Europe (and California iirc).

**Comment by pfrazee** [comment]
If there's a regulatory requirement we should do it on first tap I figure. If you have to go into settings and enable it then nobody will.

**Comment by mackuba** [comment]
> As a side-note. The app currently automatically opts-in users to this functionality. However, from a data privacy perspective this is kind of edgy since the app now per default (at least) leaks the user's IP address to 3rd parties like Google, Meta, etc. pp. So imho this feature should be opt-in and there should be a disclaimer informing the users what information is passed onto the 3rd parties when enabling the feature.

> It only contacts those services if the users tap into it and I think people generally understand that those services have to be contacted to display content from them

Uh, so I have the exact opposite request, I'm currently seeing gray rectangles in place of thumbnails until I click on the box, and I want to see the thumbnail immediately ;)

**Comment by pfrazee** [comment]
yeah something seems to be inconsistent about the metadata fetch now, I gotta look into that

**Comment by surfdude29** [comment]
> If there's a regulatory requirement we should do it on first tap I figure. If you have to go into settings and enable it then nobody will.

As well as a short explanation shown on first tap, it might be a good idea to add a section to your Privacy Policy explaining the behaviour?

https://blueskyweb.xyz/support/privacy-policy

e.g. perhaps by updating and expanding this section:

![IMG_8904](https://github.com/bluesky-social/social-app/assets/149612116/dec5cf25-add9-4de9-95ce-b7309ba0c33d)

**Comment by pfrazee** [comment]
Opened an issue on the interstitial #2338

**Comment by WolfWings** [comment]
As someone that's been active on imgur for over a decade, before giphy/tenor took over it was _THE_ place (and still is for certain styles/eras) to post/store/share animated meme images, often tailored ones or highly customized ones, as well as supporting much higher resolutions than most other options, 20MB for static images, and up to 10k on a side, thus the source of the 'intergalactic' quality image macros.

It lacks any good search tools (lack of text being added to most posts, and the tag system being too limited and itself meme-ified as a result), but folks build extensive bookmarks/folder chains on-site and share them around to catalog images, it's basically the source of "Nice Meme, yoink!" type image-macros, etc. It also has a pretty extensive 'directly import video from other sites' toolkit built right in for making clips to turn into memes/image macros, etc.

It might be a little niche, but supporting directly embedding imgur content would be convenient for a lot of the reasons folks want to be able to post videos (yes, for the memes and other short clips), if you haven't been active on imgur across that whole decade-plus existence it can be an EXTREMELY opaque culture/site to explore due to how much of it is cyclically tied to specific dates as well for when certain content does/doesn't exist, etc.

**Comment by HeyItsLollie** [comment]
Does Bluesky's YouTube embeds currently handle Playlist links?
eg: https://www.youtube.com/playlist?list=PL8rSRNukjnn4GyPiGYY_wfnv03t8E131C

I'd also like to request:
- [Bandcamp](https://bandcamp.com/) (Track & Album)
- [Sketchfab](https://sketchfab.com/)
- [Streamable](https://streamable.com/)

**Comment by HeyItsLollie** [comment]
> Notice that the embed URL is using an album ID while the share link is not. I'm not certain how we'd be able to get that or if it is possible,

Thankfully, it's all in the pages' `<meta>` tags, under `bc-page-properties`.

https://homesick89.bandcamp.com/album/exitminilp007-homesick-1000db-stare-mini-lp
`<meta name="bc-page-properties" content="{'item_type':'a','item_id':3394138826,'tralbum_page_version':0}">`

https://homesick89.bandcamp.com/track/1000db-stare
`<meta name="bc-page-properties" content="{'item_type':'t','item_id':3559172457,'tralbum_page_version':0}">`

- `item_type` can be `a` or `t` for Albums or Tracks respectively (share links include "album" or "track" in the URL)
- `item_id` is the magic number.

Additionally, embeddable player URLs are provided upfront under `og:video` \ `og:video:secure_url` and `twitter:player`

`og:video` : https://bandcamp.com/EmbeddedPlayer/v=2/album=3394138826/size=large/tracklist=false/artwork=small/
`twitter:player` : https://bandcamp.com/EmbeddedPlayer/v=2/track=3559172457/size=large/linkcol=0084B4/notracklist=true/twittercard=true/

(note: `tracklist=false` and `notracklist=true` are completely interchangeable)

This embeddable URL is easy to manipulate and customize. For example, here's an album player with large artwork, tracklist enabled, a dark background and yellow links. The order of parameters doesn't appear to matter.
https://bandcamp.com/EmbeddedPlayer/v=2/album=3541844592/size=large/artwork=large/bgcol=000000/linkcol=ffd200/tracklist=true/transparent=true/

---

> Have you seen anywhere else where these embeds are automatically taken care of?

Tumblr, Twitter, and Cohost.

I believe Tumblr uses the `og:video` URL provided above, which matches Bandcamp's "Standard" player with "Small" artwork.

![firefox - 2024-01-02 - 12-24-15](https://github.com/bluesky-social/social-app/assets/30498581/8cee57ec-2e2f-4903-90c9-68c72155971f)

Twitter's embedded player seems to differ slightly from the `twitter:player` URL when visited directly, but Cohost uses Iframely to process its embeds, and the resulting embed virtually matches Twitter. (Minor style changes, eg: Twitter has a centered Play button, both have different link colors.)

Both appear to use Bandcamp's "Artwork-only" player.

![firefox - 2024-01-02 - 13-42-40](https://github.com/bluesky-social/social-app/assets/30498581/174c32ff-58fc-4cd1-833a-3cfe899b5410)
![firefox - 2024-01-02 - 13-54-19](https://github.com/bluesky-social/social-app/assets/30498581/65022318-220a-471e-96fc-569338361e5b)

**Comment by WolfWings** [comment]
> I suppose we could always just display the gif player even before we know if it is a gif or not, but that seems like a weird UX if you're thinking you're going to be playing a gif but then it doesn't animate...

Actually a bit more digging...

...turns out they DO indicate the difference in the OpenGraph details.

https://www.opengraph.xyz/url/https%3A%2F%2Fimgur.com%2FTgdRHOd (still image)

https://www.opengraph.xyz/url/https%3A%2F%2Fimgur.com%2FBXsPWci (animated image)

Animated images always get a .gif for the twitter:image and og:image parameters, and likewise still images never do.

Avoids the API rate limiting and since opengraph is used for so much embedding details imgur exposing animated vs not via that rubrik should be usable I think? Maybe in the other issue you opened instead of this one, I'm good at tackling API nonsense but I admit I don't feel up to building the BSky app to test ideas directly in the codebase.

**Comment by YuriGal** [comment]
If this is the right place to ask for additional players  - would you consider adding YouTube Music https://music.youtube.com ?

**Comment by Zero3K** [comment]
Please add support for links such as https://packaged-media.redd.it/tajnlkezevdc1/pb/m2-res_472p.mp4?m=DASHPlaylist.mpd&v=1&e=1705950000&s=a066af59715c9eb450f69037e71c95541aac2565#t=0.

**Comment by HeyItsLollie** [comment]
I can already say there's no way Bluesky will implement embeds for direct links to video files. Why would you want to link to this file directly, instead of the Reddit post that this video originates from?

**Comment by Zero3K** [comment]
Because that is complicated. I take it that in order for video support to work, it has to be embedded from a different site which has to be allowed by the developers for the code to be added or hosted directly on one of bsky.app's servers.

**Comment by steveharoz** [comment]
I'd like to suggest a static fallback for sites like TikTok whose embed are a bit trickier?

Currently, including a link card from Tiktok just shows the root site's title, which isn't especially informative. Instead, Tiktok's oembed response ([docs here](https://developers.tiktok.com/doc/embed-videos/)) provides: 
* title
* thumbnail_width
* thumbnail_heigh
* thumbnail_URL 

How about at least using that info for a static card that links to the video?
